lib/Dialect/FIRRTL/Transforms/ExtractInstances.cpp

```diff
@@ -0,0 +1,409 @@
+  if (!anythingChanged)
+    markAllAnalysesPreserved();
+  markAnalysesPreserved<InstanceGraph>();
+}
+
+static bool isAnnoInteresting(Annotation anno) {
+  return anno.isClass(extractBlackBoxAnnoClass);
+}
+
+/// Gather the modules and instances annotated to be moved by this pass. This
+/// populates the corresponding lists and maps of the pass.
+void ExtractInstancesPass::collectAnnos() {
+  CircuitOp circuit = getOperation();
+
+  // Gather the annotations on modules. These complement the later per-instance
+  // annotations.
+  circuit.walk([&](FModuleLike module) {
+    AnnotationSet::removeAnnotations(module, [&](Annotation anno) {
+      if (anno.isClass(dutAnnoClass)) {
+        LLVM_DEBUG(llvm::dbgs()
+                   << "Marking DUT `" << module.moduleName() << "`\n");
+        dutModules.insert(module);
+        return false; // other passes may rely on this anno; keep it
+      }
+      if (!isAnnoInteresting(anno))
+        return false;
+      LLVM_DEBUG(llvm::dbgs() << "Annotated module `" << module.moduleName()
+                              << "`:\n  " << anno.getDict() << "\n");
+      annotatedModules[module].push_back(anno);
+      return true;
+    });
+    return WalkResult::skip();
+  });
+
+  // Gather the annotations on instances to be extracted.
+  circuit.walk([&](InstanceOp inst) {
+    SmallVector<Annotation, 1> instAnnos;
+    Operation *module = instanceGraph->getReferencedModule(inst);
+
+    // Module-level annotations.
+    auto it = annotatedModules.find(module);
+    if (it != annotatedModules.end())
+      instAnnos.append(it->second);
+
+    // Instance-level annotations.
+    AnnotationSet::removeAnnotations(inst, [&](Annotation anno) {
+      if (!isAnnoInteresting(anno))
+        return false;
+      LLVM_DEBUG(llvm::dbgs() << "Annotated instance `" << inst.name()
+                              << "`:\n  " << anno.getDict() << "\n");
+      instAnnos.push_back(anno);
+      return true;
+    });
+
+    // No need to do anything about unannotated instances.
+    if (instAnnos.empty())
+      return;
+
+    // Ensure there are no conflicting annotations.
+    if (instAnnos.size() > 1) {
+      auto d = inst.emitError("multiple extraction annotations on instance `")
+               << inst.name() << "`";
+      d.attachNote(inst.getLoc()) << "instance has the following annotations, "
+                                     "but at most one is allowed:";
+      for (auto anno : instAnnos)
+        d.attachNote(inst.getLoc()) << anno.getDict();
+      anyFailures = true;
+      return;
+    }
+
+    // Process the annotation.
+    collectAnno(inst, instAnnos[0]);
+  });
+
+  // Propagate the DUT marking to all arbitrarily nested submodules of the DUT.
+  LLVM_DEBUG(llvm::dbgs() << "Marking DUT hierarchy\n");
+  SmallVector<InstanceGraphNode *> worklist;
+  for (Operation *op : dutModules)
+    if (auto module = dyn_cast<hw::HWModuleLike>(op))
+      worklist.push_back(instanceGraph->lookup(module));
```

Comment:
The PR could avoid the `dyn_cast` and blindly cast here given earlier code.

```suggestion
    auto module = cast<hw::HWModuleLike>(op);
    worklist.push_back(instanceGraph->lookup(module));
```

Reply:
Cool, I missed that :+1: